---
title: "R Notebook"
output: html_notebook
---

## Loading Packages
```{r, error = F, message = F, warning = F}
library(ggsci)
library(DESeq2)
library(ggpubr)
library(viridis)
library(scales)
library(rtracklayer)
library(GenomicRanges)
library(BiocParallel)
library(tiff)
library(RColorBrewer)
library(ComplexHeatmap)
library(circlize)
library(BRGenomics)
library(extrafont)
library(plyr)
loadfonts()
library(FSA)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(tidyverse)
library(karyoploteR)
library(patchwork)
library(biomaRt)
source("~/Google Drive/R/customPackages/jj_functions.R")
source("~/Google Drive/github_repos/browser_plot.R/browser_plot.R")
```

```{r}
maplot.jj <- function(res, padj_cutoff = 0.01, l2fc_cutoff = 0) {
    # This function takes a DESeq2 Results object and plots a pretty MA plot.
    # res = the results object
    # padj_cutoff = significance cutoff (dbl)
    # l2fc_cutoff = log2 fold change cutoff (dbl)
    
    resdf <- as.data.frame(res)
    resdf$name <- row.names(resdf)
    if(!(sum(is.na(resdf$padj)) == 0)){
    resdf[is.na(resdf$padj),]$padj <- 1}
    resdf <-
        drop_na(mutate(resdf, "class" = if_else(
            padj < padj_cutoff,
            if_else(log2FoldChange > 0, "Activated", "Repressed"),
            "Unchanged"
        )))
    
    resdf$class[which(abs(resdf$log2FoldChange) < l2fc_cutoff)] <- "Unchanged"
    
    numAct <- nrow(filter(resdf, class == "Activated"))
    numRep <- nrow(filter(resdf, class == "Repressed"))
    numUch <- nrow(filter(resdf, class == "Unchanged"))
    max_y <- max(resdf$log2FoldChange)
    min_y <- min(resdf$log2FoldChange)
    max_x <- max(resdf$baseMean)
    min_x <- min(resdf$baseMean)
    x_lim_left <- 10^floor(log10(min_x))
    x_lim_right <- 10^ceiling(log10(max_x))
    y_lim_top <- round_any(max_y, 2, ceiling)
    y_lim_bottom <- round_any(min_y, 2, floor)
    
    lseq <- function(from, to, length.out) {
        # logarithmic spaced sequence
        # blatantly stolen from library("emdbook"), because need only this
        exp(seq(log(from), log(to), length.out = length.out))
    }
    
    breaks_x <- lseq(
        from = x_lim_left,
        to = x_lim_right,
        length.out = (log10(x_lim_right) - log10(x_lim_left) + 1)
    )
    
    breaks_y <- seq(y_lim_bottom, y_lim_top, 2)
    
    p1 <-
        ggplot(resdf, aes(x = baseMean, y = log2FoldChange, color = class)) +
        geom_point(
                   size = 0.75,
                   show.legend = F) +
        xlab('log10 Mean Expression') +
        ylab('log2 Fold Change') +
        scale_x_log10(
            limits = c(x_lim_left, x_lim_right),
            breaks = breaks_x,
            labels = round(log10(breaks_x)),
            expand = c(0,0)
        ) +
        scale_y_continuous(
            limits = c(y_lim_bottom, y_lim_top),
            expand = c(0,0),
            breaks = breaks_y,
            labels = breaks_y
        )+
        scale_color_manual(
            values = c("Activated" = "#BB0021", 
                       "Repressed" = "#3B4992", 
                       "Unchanged" = "gray"),
        )+
        geom_hline(yintercept = 0,
                   size = 0.5,
                   alpha = 1) +
        annotate(geom = 'text', 
                 label = numAct, 
                 x = min_x, 
                 y = max_y, 
                 color = "#BB0021", 
                 hjust = -0.25, 
                 vjust = 0.5)+
        annotate(geom = 'text', 
                 label = numRep, 
                 x = min_x, 
                 y = min_y, 
                 color = "#3B4992", 
                 hjust = -0.25, 
                 vjust = -0.5)+
        ggtheme.jj() +
        theme() 
    
    if(l2fc_cutoff > 0){
        p1 <- p1+
            geom_hline(yintercept = l2fc_cutoff,
                       size = 0.5,
                       color = "grey",
                       linetype = "dashed")+
            geom_hline(yintercept = -l2fc_cutoff,
                       size = 0.5,
                       color = "grey",
                       linetype = "dashed")
    }
    return(p1)
}
```

## Importing ATAC-seq data
```{r}
ATAC.lst <- import.bw.list("bw/merged_bw/", stranded = FALSE)

# Normalizing each by reads per million
ATAC.lst <- lapply(ATAC.lst, RPMnorm)

# Exporting RPM normed bw files
dir.create("bw/merged_RPM")
lapply(
  X = names(ATAC.lst),
  FUN = export.bw.lst,
  path = "bw/merged_RPM/",
  GRanges.lst = ATAC.lst
)
```

## Importing annotation data
```{r}
genes.gr <- 
    transcripts(
      TxDb.Hsapiens.UCSC.hg38.knownGene,
      columns = c("GENEID", "TXNAME")
      )
tes.gr <- import("hg38_rmOut.gtf", format = "gtf")

ensembl = useMart("ensembl")
ensembl = useDataset("hsapiens_gene_ensembl", mart=ensembl)
genes_id.df <- 
    getBM(attributes = c("chromosome_name", "start_position", "end_position", "strand", "hgnc_symbol", "ensembl_gene_id", "ensembl_transcript_id"), mart = ensembl)

colnames(genes_id.df) <- c("chr", "start", "end", "strand", "name")

genes_id.df$strand <- ifelse(genes_id.df$strand > 0, "+", "-")
genes_id.df$chr <- paste0("chr", genes_id.df$chr)

genes_id.gr <- makeGRangesFromDataFrame(genes_id.df, keep.extra.columns = T)


```

## Runnning DE analysis on ATAC-seq peaks
```{r}
# Reading in counts matrix
peak_counts.mat <- read.table(
  "counts/counts_peaks.tsv",
  row.names = NULL
)

# Defining column names
colnames(peak_counts.mat) <- c(
  "peak_name",
  "JH10_low_MMPi_Rep2",
  "JH11_low_MMPi_Rep3",
  "JH12_low_MMPi_Rep4",
  "JH1_high_MMPi_Rep1",
  "JH2_high_MMPi_Rep2",
  "JH3_high_MMPi_Rep3",
  "JH4_high_MMPi_Rep4",
  "JH5_medium_MMPi_Rep1",
  "JH6_medium_MMPi_Rep2",
  "JH7_medium_MMPi_Rep3",
  "JH8_medium_MMPi_Rep4",
  "JH9_low_MMPi_Rep1"
)

# Formatting df for DESeq
row.names(peak_counts.mat) <- peak_counts.mat$peak_name

peak_counts.mat <- peak_counts.mat[,-1]

# Making metadata object
col.data <- data.frame(row.names = colnames(peak_counts.mat),
                       cond = c("low", 
                                "low", 
                                "low", 
                                "high", 
                                "high",
                                "high", 
                                "high", 
                                "medium",
                                "medium",
                                "medium",
                                "medium",
                                "low"
                                ),
                       rep = c(2, 3, 4, 
                               1, 2, 3, 4,
                               1, 2, 3, 4,
                               1),
                       sample = c("JH10",
                                  "JH11",
                                  "JH12",
                                  "JH1", 
                                  "JH2",
                                  "JH3", 
                                  "JH4", 
                                  "JH5", 
                                  "JH6", 
                                  "JH7", 
                                  "JH8", 
                                  "JH9")
                       )



# Making DESeq object
dds.peaks <- DESeqDataSetFromMatrix(
  peak_counts.mat, col.data, design = ~cond
)

# Factoring treatment levels
dds.peaks$cond <- factor(dds.peaks$cond, levels = c("low", "medium", "high"))

# Running DESeq
dds.peaks <- DESeq(dds.peaks)

dir.create("MA")
dir.create("results")
dir.create("bed")

# Comparing high to low
res.peaks.highvslow <- results(dds.peaks, contrast= c("cond", "high", "low"))

high_vs_low_ma.plot <- maplot.jj(res.peaks.highvslow, 0.1) + 
  ggtitle("High vs. Low")

res.peaks.highvslow.df <- as.data.frame(res.peaks.highvslow)
write.table(res.peaks.highvslow.df,
            "results/peaks_high_vs_low.tsv", 
            quote = F, 
            sep = "\t")

res.peaks.highvslow.df <- 
  res.peaks.highvslow.df[res.peaks.highvslow.df$padj < 0.1, ]

res.peaks.highvslow.df$coord <- row.names(res.peaks.highvslow.df)

res.peaks.highvslow.df <- 
  separate(res.peaks.highvslow.df, coord, into = c("chr", "start", "end"), sep = "\\.")

res.peaks.highvslow.gr <- makeGRangesFromDataFrame(
  res.peaks.highvslow.df
)

res.peaks.highvslow.gr$l2FC <- res.peaks.highvslow.df$log2FoldChange

res.peaks.highvslow.gr <- 
  res.peaks.highvslow.gr[-which(is.na(nearest(res.peaks.highvslow.gr, genes.gr))),]

res.peaks.highvslow.gr$nearest_tx <- 
  genes.gr[nearest(res.peaks.highvslow.gr, genes.gr),]$tx_name

res.peaks.highvslow.bed <- as.data.frame(res.peaks.highvslow.gr)
row.names(res.peaks.highvslow.bed) <- NULL
res.peaks.highvslow.bed <- res.peaks.highvslow.bed[, c(1,2,3,7,6,5)]

write.table(res.peaks.highvslow.bed, "bed/peaks_high_vs_low_annotated.bed",
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

# Comparing medium to low
res.peaks.mediumvslow <- results(dds.peaks, contrast= c("cond", "medium", "low"))


medium_vs_low_ma.plot <- maplot.jj(res.peaks.mediumvslow, 0.1) + 
  ggtitle("Medium vs. Low")

res.peaks.mediumvslow.df <- as.data.frame(res.peaks.mediumvslow)
write.table(res.peaks.mediumvslow.df,
            "results/peaks_medium_vs_low.tsv", 
            quote = F, 
            sep = "\t")

res.peaks.mediumvslow.df <- 
  drop_na(res.peaks.mediumvslow.df[res.peaks.mediumvslow.df$padj < 0.1, ])

res.peaks.mediumvslow.df$coord <- row.names(res.peaks.mediumvslow.df)

res.peaks.mediumvslow.df <- 
  separate(res.peaks.mediumvslow.df, coord, into = c("chr", "start", "end"), sep = "\\.")

res.peaks.mediumvslow.gr <- makeGRangesFromDataFrame(
  res.peaks.mediumvslow.df
)

res.peaks.mediumvslow.gr$l2FC <- res.peaks.mediumvslow.df$log2FoldChange

res.peaks.mediumvslow.gr <- 
  res.peaks.mediumvslow.gr[-which(is.na(nearest(res.peaks.mediumvslow.gr, genes.gr))),]

res.peaks.mediumvslow.gr$nearest_tx <- 
  genes.gr[nearest(res.peaks.mediumvslow.gr, genes.gr),]$tx_name

res.peaks.mediumvslow.bed <- as.data.frame(res.peaks.mediumvslow.gr)
row.names(res.peaks.mediumvslow.bed) <- NULL
res.peaks.mediumvslow.bed <- res.peaks.mediumvslow.bed[, c(1,2,3,7,6,5)]

write.table(res.peaks.mediumvslow.bed, "bed/peaks_medium_vs_low_annotated.bed",
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

# Comparing high to medium
res.peaks.highvsmedium <- results(dds.peaks, contrast= c("cond", "high", "medium"))

high_vs_medium_ma.plot <- maplot.jj(res.peaks.highvsmedium, 0.1) + 
  ggtitle("High vs. Medium")

res.peaks.highvsmedium.df <- as.data.frame(res.peaks.highvsmedium)
write.table(res.peaks.highvsmedium.df,
            "results/peaks_high_vs_medium.tsv", 
            quote = F, 
            sep = "\t")

res.peaks.highvsmedium.df <- 
  drop_na(res.peaks.highvsmedium.df[res.peaks.highvsmedium.df$padj < 0.1, ])

res.peaks.highvsmedium.df$coord <- row.names(res.peaks.highvsmedium.df)

res.peaks.highvsmedium.df <- 
  separate(res.peaks.highvsmedium.df, coord, into = c("chr", "start", "end"), sep = "\\.")

res.peaks.highvsmedium.gr <- makeGRangesFromDataFrame(
  res.peaks.highvsmedium.df
)

res.peaks.highvsmedium.gr$l2FC <- res.peaks.highvsmedium.df$log2FoldChange

res.peaks.highvsmedium.gr <- 
  res.peaks.highvsmedium.gr[-which(is.na(nearest(res.peaks.highvsmedium.gr, genes.gr))),]

res.peaks.highvsmedium.gr$nearest_tx <- 
  genes.gr[nearest(res.peaks.highvsmedium.gr, genes.gr),]$tx_name

res.peaks.highvsmedium.bed <- as.data.frame(res.peaks.highvsmedium.gr)
row.names(res.peaks.highvsmedium.bed) <- NULL
res.peaks.highvsmedium.bed <- res.peaks.highvsmedium.bed[, c(1,2,3,7,6,5)]

write.table(res.peaks.highvsmedium.bed, "bed/peaks_high_vs_medium_annotated.bed",
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

high_vs_low_ma.plot + high_vs_medium_ma.plot + medium_vs_low_ma.plot

ggsave("MA/all_MA.pdf", width = 5, height = 2, useDingbats = FALSE)

# PCA plot
dds.peaks.rld <- rlog(dds.peaks, blind = FALSE)

peaks.pca <-
   plotPCA(dds.peaks.rld,
           intgroup = c("sample"),
           returnData = TRUE)

# Getting % Var explained by each PC
peaks_pcavar.dbl <- round(100 * attr(peaks.pca, "percentVar"))

peaks.pca$cond <- ifelse(grepl("high", peaks.pca$name), 
                         "high", 
                         ifelse(grepl("medium", peaks.pca$name), 
                                "medium", 
                                "low")
                         )
peaks.pca <- separate(peaks.pca, name, c(NA, NA, NA, "Rep"))

peaks.pca$cond <- factor(peaks.pca$cond, levels = c("low", "medium", "high"))

# Plotting
dir.create("PCA")

ggsave(
  "PCA/peaks_PCA.pdf",
  (ggplot(peaks.pca, aes(PC1, PC2, color = cond)) +
    geom_point(size = 1, aes()) +
    xlab(paste0("PC1: ", peaks_pcavar.dbl[1], "% var.")) +
    ylab(paste0("PC2: ", peaks_pcavar.dbl[2], "% var.")) +
    ggtheme.jj() +
     scale_y_continuous(limits = c(-10, 10), breaks = c(-10,0,10), expand = c(0,0))+
     scale_x_continuous(limits = c(-10, 10), breaks = c(-10, 0, 10), expand = c(0,0)) +
     scale_color_aaas()+
     theme(legend.title = element_blank())
  ),
  width = 3,
  height = 2
)

# Plotting metaprofile around peaks
up.gr <- res.peaks.highvslow.gr[res.peaks.highvslow.gr$l2FC > 0, ]
down.gr <- res.peaks.highvslow.gr[res.peaks.highvslow.gr$l2FC < 0, ]

# Centering peaks and extending to +/- 500
up.gr$center <- as.integer(start(up.gr) + ((end(up.gr) - start(up.gr)) / 2))
start(up.gr) <- up.gr$center - 1000
end(up.gr) <- up.gr$center + 1000

down.gr$center <- as.integer(start(down.gr) + ((end(down.gr) - start(down.gr)) / 2))
start(down.gr) <- down.gr$center - 1000
end(down.gr) <- down.gr$center + 1000


up.mat <- metaSubsample(
  ATAC.lst,
  up.gr,
  first.output.xval = -1000,
  binsize = 10,
  expand_ranges = TRUE
)

up.mat$class <- "up"

down.mat <- metaSubsample(
  ATAC.lst,
  down.gr,
  first.output.xval = -1000,
  binsize = 10,
  expand_ranges = TRUE
)

down.mat$class <- "down"

all.mat <- rbind(up.mat, down.mat)

all.mat$sample.name <- factor(all.mat$sample.name, levels = c("Low", "Medium", "High"))

ggMetaplot(down.mat) +
  scale_fill_aaas() + 
  scale_color_aaas() + 
  scale_x_continuous(limits = c(-1000, 1000), breaks = c(-1000, -500, 0, 500, 1000),
                     expand = c(0,0)) +
  scale_y_continuous(limits = c(0, 0.0075), breaks = c(0, 0.0025, 0.0050, 0.0075),
                     expand = c(0,0)) + 
  theme(legend.title = element_blank()) +
  xlab("Distance to peak center (bp)")

dir.create("metaplots")
ggsave("metaplots/DE_peaks_ATAC_signal_metaplot.pdf", width = 4, height = 2)

ggsave("metaplots/down_peaks_ATAC_signal_downOnlyRescaled.pdf", width = 2.5, height = 1.4)


```


## Making karyoplot of DE peak locations
```{r}
pdf("karyotype/up_peaks_karyoplot.pdf", width = 4, height = 4)
kp <- plotKaryotype(genome="hg38", ideogram.plotter = NULL)
kpAddCytobandsAsLine(kp)
kpPlotRegions(kp, data=up.gr)
dev.off()

pdf("karyotype/down_peaks_karyoplot.pdf", width = 4, height = 4)
kp <- plotKaryotype(genome="hg38", ideogram.plotter = NULL)
kpAddCytobandsAsLine(kp)
kpPlotRegions(kp, data=down.gr)
dev.off()

```


# Exporting bed files of DE peaks
```{r}
up.gr$score <- up.gr$l2FC
up.gr$name <- up.gr$nearest_tx
export.bed(up.gr, "bed/up_peaks.bed")

down.gr$score <- down.gr$l2FC
down.gr$name <- down.gr$nearest_tx
export.bed(up.gr, "bed/up_peaks.bed")

```


# Plotting enrichments in DE peaks
# Calculated outside of R by "TE-analysis_Shuffle_bed.pl", https://github.com/4ureliek/TEanalysis
```{r}
up_enrich.mat <- 
    read.table(
      "enrich/up_shuf/up_shuf.tsv",
        skip = 14,
        col.names = c("Rclass", "Rfam", "Rname", "obs_hits", "perc_obs",
                   "obs_tot_hits", "no_trials", "exp_hits", "exp_sd", 
                   "perc_exp", "exp_tot_hits", "rank", "perm_pval",
                   "sig1", "binomial_prob", "binomial_95CI", "binom_pval","sig2"),
        row.names = NULL
        )

up_enrich.mat$l2OE <- log2((up_enrich.mat$obs_hits + 1) / (up_enrich.mat$exp_hits + 1))
up_enrich.mat <- up_enrich.mat[!up_enrich.mat$Rname == "tot", ]
up_enrich.mat <- up_enrich.mat[!up_enrich.mat$sig2 == "na", ]

up_enrich.plot <- ggplot(up_enrich.mat, aes(x = l2OE, y = obs_hits, color = sig2)) +
  geom_point() +
  scale_color_manual(
    values = c(
      "ns" = "#BBBBBB",
      "*" = "#AA3377",
      "**" = "#4477AA",
      "***" = "#228833"
    ),
    labels = c(
      "ns" = "ns",
      "*" = "p < 0.1",
      "**" = "p < 0.01",
      "***" = "p < 0.001"
    ),
    name = "Binomial Test"
  ) +
  ggtheme.jj() +
  xlab("log2(obs./exp.)") +
  ylab("# Obs.") +
  scale_x_continuous(limits = c(-4, 2), expand = c(0, 0)) +
  scale_y_continuous(limits = c(0, 60), expand = c(0, 0)) +
  ggtitle("Up Peaks")

down_enrich.mat <- 
    read.table(
        "enrich/down_shuf/down_shuf.tsv",
        skip = 14,
        col.names = c("Rclass", "Rfam", "Rname", "obs_hits", "perc_obs",
                   "obs_tot_hits", "no_trials", "exp_hits", "exp_sd", 
                   "perc_exp", "exp_tot_hits", "rank", "perm_pval",
                   "sig1", "binomial_prob", "binomial_95CI", "binom_pval","sig2"),
        row.names = NULL
        )

down_enrich.mat$l2OE <- log2((down_enrich.mat$obs_hits + 1) / (down_enrich.mat$exp_hits + 1))
down_enrich.mat <- down_enrich.mat[!down_enrich.mat$Rname == "tot", ]
down_enrich.mat <- down_enrich.mat[!down_enrich.mat$sig2 == "na", ]

down_enrich.plot <- ggplot(down_enrich.mat, aes(x = l2OE, y = obs_hits, color = sig2)) +
  geom_point() +
  scale_color_manual(
    values = c(
      "ns" = "#BBBBBB",
      "*" = "#AA3377",
      "**" = "#4477AA",
      "***" = "#228833"
    ),
    labels = c(
      "ns" = "ns",
      "*" = "p < 0.1",
      "**" = "p < 0.01",
      "***" = "p < 0.001"
    ),
    name = "Binomial Test"
  ) +
  ggtheme.jj() +
  xlab("log2(obs./exp.)") +
  ylab("# Obs.") +
  scale_x_continuous(limits = c(-5, 5), expand = c(0, 0)) +
  scale_y_continuous(limits = c(0, 250), expand = c(0, 0)) +
  ggtitle("Down Peaks")

up_enrich.plot + down_enrich.plot + plot_layout(guides = 'collect')

ggsave("enrich/TE_enrichment.pdf", width = 5, height = 2, useDingbats = FALSE)
```


# Plotting distributions of TEs enriched in Down
```{r}
AluYj4.gr <- tes.gr[tes.gr$transcript_id == "AluYj4"]

pdf("karyotype/AluYj4_karyoplot.pdf", width = 4, height = 4)
kp <- plotKaryotype(genome="hg38", ideogram.plotter = NULL)
kpAddCytobandsAsLine(kp)
kpPlotRegions(kp, data=AluYj4.gr)
dev.off()


ALR_Alpha.gr <- tes.gr[tes.gr$gene_id == "ALR/Alpha"]

pdf("karyotype/ALR_Alpha_karyoplot.pdf", width = 4, height = 4)
kp <- plotKaryotype(genome="hg38", ideogram.plotter = NULL)
kpAddCytobandsAsLine(kp)
kpPlotRegions(kp, data=ALR_Alpha.gr)
dev.off()


```
# Browser shots
```{r}
ATAC.lst <- ATAC.lst[c(2,3,1)]
HDAC3 <- genes_id.gr[genes_id.gr$name == "HDAC3"]
CBX5 <- genes_id.gr[genes_id.gr$name == "CBX5"]
TRIM44 <- genes_id.gr[genes_id.gr$name == "TRIM44"]
TERT <- genes_id.gr[genes_id.gr$name == "TERT"]
COL23A1 <- genes_id.gr[genes_id.gr$name == "COL23A1"]
MUC2 <- genes_id.gr[genes_id.gr$name == "MUC2"]

HDAC3_browser <- 
    browser_plotter(
        HDAC3, 
        list("ATACseq" = ATAC.lst), 
        binsize = 25, 
        pad_right = 5000)

CBX5_browser <- 
    browser_plotter(
        CBX5, 
        list("ATACseq" = ATAC.lst), 
        binsize = 50, 
        pad_right = 5000)

TRIM44_browser <- 
    browser_plotter(
        TRIM44, 
        list("ATACseq" = ATAC.lst), 
        binsize = 100, 
        pad_left = 5000)

TERT_browser <- 
    browser_plotter(
        TERT, 
        list("ATACseq" = ATAC.lst), 
        binsize = 50, 
        pad_right = 5000)

COL23A1_browser <- 
    browser_plotter(
        COL23A1, 
        list("ATACseq" = ATAC.lst), 
        binsize = 100, 
        pad_reft = 5000)

MUC2_browser <- 
    browser_plotter(
        MUC2, 
        list("ATACseq" = ATAC.lst), 
        binsize = 25, 
        pad_left = 5000)

dir.create("browser_shots")

pdf("browser_shots/MUC2.pdf", width = 2, height = 3)
grid.draw(MUC2_browser)
dev.off


```

